projects/client/RabbitMQ.Client/src/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -43,14 +43,16 @@ class WorkPool
             readonly ConcurrentQueue<Work> _workQueue;
             readonly CancellationTokenSource _tokenSource;
             readonly ModelBase _model;
-            readonly SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+            CancellationTokenRegistration _tokenRegistration;
+            volatile TaskCompletionSource<bool> _syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Using `TaskCreationOptions.RunContinuationsAsynchronously` because that prevents issues like running the continuation of the loop in the worker on the thread that is setting the TCS with `TrySetResult` (that would be the one enqueuing)